src/Controller/ApiV1Controller.php

```diff
@@ -0,0 +1,162 @@
+
+        if (!Configure::read('Search.enabled')) {
+            $this->render('search_disabled');
+            return;
+        }
+
+        /* Apply search criteria and sort */
+        $search = new SentencesSearchForm();
+        $search->setData($this->request->getQueryParams());
+
+        /* Control input */
+        if ($search->generateRandomSeedIfNeeded()) {
+            return $this->redirect(Router::url($search->getData()));
+        }
+        $search->checkUnwantedCombinations();
+
+        $limit = CurrentUser::getSetting('sentences_per_page');
+        $sphinx = $search->asSphinx();
+        $sphinx['page'] = $this->request->query('page');
+//        $sphinx['limit'] = $limit;
+
+        // TODO: improve on hard-coded limit to the number of API results
+        $sphinx['limit'] = 100; // Limit the API to returning 100 entries.
+
+        $model = 'Sentences';
+        $query = $this->Sentences
+            ->find('hideFields')
+            ->find('withSphinx')
+            ->find('nativeMarker')
+            ->find('filteredTranslations', [
+                'translationLang' => $search->getData('to'),
+            ])
+            ->select($this->Sentences->fields())
+            ->contain($this->Sentences->contain(['translations' => true]));
+
+        $this->paginate = [
+            'limit' => $limit,
+            'sphinx' => $sphinx,
+        ];
+
+        try {
+            $results = $this->paginate($query);
+            $real_total = $this->Sentences->getRealTotal();
+            $results = $this->Sentences->addHighlightMarkers($results);
+            $this->set(compact('results', 'real_total'));
+        } catch (Exception $e) {
+            $syntax_error = strpos($e->getMessage(), 'syntax error,') !== FALSE;
+            if ($syntax_error) {
+                $this->set('syntax_error', true);
+            } else {
+                $this->loadComponent('Error');
+                $error_code = $this->Error->traceError('Search error: ' . $e->getMessage());
+                $this->set('error_code', $error_code);
+            }
+        }
+
+        $strippedQuery = preg_replace('/"|=/', '', $search->getData('query'));
+        $this->loadModel('Vocabulary');
+        $vocabulary = $this->Vocabulary->findByText($strippedQuery);
+
+        $searchableLists = $search->getSearchableLists(CurrentUser::get('id'));
+        $ignored = $search->getIgnoredFields();
+
+        $this->set($search->getData());
+        $this->set(compact('ignored', 'searchableLists', 'vocabulary'));
+        $this->set(
+            'is_advanced_search',
+            !is_null($this->request->getQuery('trans_to'))
+        );
+
+        $ids = array();
+        foreach ($results as $sentence) {
+            $ids[] = $sentence->id;
+        }
+    
+        $ids_json = json_encode($ids);
+        $ids_json = '{ "ids" : ' . $ids_json . ' }';
+        return $this->getResponse()
+            ->withStringBody($ids_json);
+    }
```

Comment:
This function has a lot of unnecessary code and I can't deny I would be sad to merge it as is. Below my take on how to trim this down to the essential. Please try it out and commit the code if that works fine for you :)

One thing I should note: I'm not exactly sure why you chose to return only the sentence ids rather than the sentence objects. My version of the code returns the whole object because I can't think of a practical use case where an app that consumes this API will need only the ids. Most likely, the API consumer will want the sentence data too, and if it only has the ids, it will be forced to do a whole bunch of `GET /sentence/{id}` after the `GET /search`. Better serve all this data in one call, in my opinion.

```
    public function search()
    {
        $this->loadModel('Sentences');

        $search = new SentencesSearchForm();
        $search->setData($this->request->getQueryParams());

        $limit = 100;
        $sphinx = $search->asSphinx();
        $sphinx['page'] = $this->request->query('page');
        $sphinx['limit'] = $limit;

        $query = $this->Sentences
            ->find('hideFields')
            ->find('withSphinx')
            ->find('filteredTranslations', [
                'translationLang' => $search->getData('to'),
            ])
            ->select($this->Sentences->fields())
            ->contain($this->Sentences->contain(['translations' => true]));

        $this->paginate = [
            'limit' => $limit,
            'sphinx' => $sphinx,
        ];

        try {
            $results = $this->paginate($query);
        } catch (Exception $e) {
            throw new \Cake\Http\Exception\BadRequestException();
        }
    
        $json = json_encode($results);
        return $this->response
            ->withType('application/json')
            ->withStringBody($json);
    }
```

Reply:
That new version looks great! I think you're right that it is better to return the complete sentence objects, otherwise the return results will just create hundreds of other calls. I think my thinking was that if there are many ids, and each sentence had many translations, then getting every id and every sentence detail in one call would cause performance or memory problems, which would be avoided by getting smaller chunks. Nonetheless, I think on balance it is better to start off with getting the sentence for each id, and if we do find there are performance/memory problems then we can look into the alternative.

I can see the secret here is to write crappy code (not that I have to try), and someone who know the code base much better than me will rewrite it as it should be written. ;-)